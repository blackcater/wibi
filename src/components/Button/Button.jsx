import VueTypes from 'vue-types';
import { COLORS, SIZES } from 'utils/constant';
import prefix, { defaultClassPrefix } from 'utils/prefix';
import SafeAnchor from 'components/SafeAnchor';

const CLASS_PREFIX = 'btn';

export default {
  name: 'Button',

  props: {
    appearance: VueTypes.oneOf([
      'default',
      'primary',
      'link',
      'subtle',
      'ghost',
    ]).def('default'),
    color: VueTypes.oneOf(COLORS),
    size: VueTypes.oneOf(SIZES),
    active: VueTypes.bool.def(false),
    disabled: VueTypes.bool.def(false),
    loading: VueTypes.bool.def(false),
    block: VueTypes.bool.def(false),
    href: VueTypes.string,
    classPrefix: VueTypes.string.def(defaultClassPrefix(CLASS_PREFIX)),
    componentClass: VueTypes.oneOfType([VueTypes.string, VueTypes.object]).def(
      'button'
    ),
  },

  computed: {
    classes() {
      const addPrefix = prefix(this.classPrefix);

      return [
        this.classPrefix,
        addPrefix(this.appearance),
        {
          [addPrefix(this.color)]: true,
          [addPrefix(this.size)]: true,
          [addPrefix('active')]: this.active,
          [addPrefix('disabled')]: this.disabled,
          [addPrefix('loading')]: this.loading,
          [addPrefix('block')]: this.block,
        },
      ];
    },
  },

  render() {
    const spin = <span class={prefix(this.classPrefix, 'spin')} />;

    if (this.href) {
      const anchorData = {
        class: this.classes,
        props: {
          ...this.$attrs,
          href: this.href,
          disabled: this.disabled,
          role: 'button',
        },
        on: { ...this.$listeners },
      };

      return (
        <SafeAnchor {...anchorData}>
          {this.loading && spin}
          {this.$slots.default}
        </SafeAnchor>
      );
    }

    const Component = this.componentClass;
    const btnData = {
      class: this.classes,
      props: {
        ...this.$attrs,
        disabled: this.disabled,
      },
      on: { ...this.$listeners },
    };

    // https://stackoverflow.com/questions/41904199/whats-the-point-of-button-type-button
    if (Component === 'button') {
      btnData.props.type = this.$attrs.type || 'button';
    }

    return (
      <Component {...btnData}>
        {this.loading && spin}
        {this.$slots.default}
      </Component>
    );
  },
};
