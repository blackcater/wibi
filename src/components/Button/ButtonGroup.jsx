import VueTypes from 'vue-types';
import prefix, { defaultClassPrefix } from 'utils/prefix';
import { SIZES } from 'utils/constant';

const CLASS_PREFIX = 'btn-group';

export default {
  name: 'ButtonGroup',

  props: {
    size: VueTypes.oneOfType(SIZES),
    vertical: VueTypes.bool.def(false),
    justified: VueTypes.bool.def(false),
    block: VueTypes.bool.def(false),
    classPrefix: VueTypes.string.def(defaultClassPrefix(CLASS_PREFIX)),
  },

  computed: {
    classes() {
      const addPrefix = prefix(this.classPrefix);

      return [
        this.classPrefix,
        {
          [addPrefix('vertical')]: this.vertical,
          [addPrefix('justified')]: this.justified,
          [addPrefix('block')]: this.block,
        },
      ];
    },
  },

  render() {
    console.dir(this.$slots.default);

    return null;
  },
};
