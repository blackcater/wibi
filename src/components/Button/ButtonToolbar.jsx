import VueTypes from 'vue-types';
import { defaultClassPrefix } from 'utils/prefix';

const CLASS_PREFIX = 'btn-toolbar';

export default {
  name: 'ButtonToolbar',

  props: {
    classPrefix: VueTypes.string.def(defaultClassPrefix(CLASS_PREFIX)),
  },

  render() {
    const btnToolbarData = {
      class: this.classPrefix,
      attrs: {
        role: 'toolbar',
        ...this.$attrs,
      },
      on: this.$listeners,
    };

    return <div {...btnToolbarData}>{this.$slots.default}</div>;
  },
};
