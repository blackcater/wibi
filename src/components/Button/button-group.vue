<template>
  <button/>
</template>

<script>
export default {
  name: 'ButtonGroup',

  props: {},

  data() {
    return {};
  },

  methods: {},
};
</script>
