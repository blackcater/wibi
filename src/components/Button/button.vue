<template>
  <button
    v-if="!href"
    :class="classes"
    :disabled="disabled"
    @click="_onClick"
  >
    <span :class="addPrefix('spin')"/>
    <slot/>
  </button>
  <safe-anchor
    v-else
    :class="classes"
    :href="href"
    @click="_onClick"
  >
    <span :class="addPrefix('spin')"/>
    <slot/>
  </safe-anchor>
</template>

<script>
import SafeAnchor from 'components/SafeAnchor';
import cls from 'classnames';
import { validateSize, validateColor, validateEnum } from 'utils/props';
import prefixMixin from 'mixins/prefixMixin';

const CLASS_PREFIX = 'btn';

export default {
  name: 'Button',

  components: { SafeAnchor },

  props: {
    active: Boolean,
    disabled: Boolean,
    loading: Boolean,
    block: Boolean,
    size: {
      validator: validateSize,
      default: 'md',
    },
    type: {
      validator: validateEnum([
        'default',
        'primary',
        'link',
        'subtle',
        'ghost',
      ]),
      default: 'default',
    },
    color: { validator: validateColor },
    href: String,
  },

  mixins: [prefixMixin(CLASS_PREFIX)],

  data() {
    return {};
  },

  computed: {
    classes() {
      return cls([
        this.classPrefix,
        this.addPrefix(this.type),
        {
          [this.addPrefix(this.color)]: this.color,
          [this.addPrefix(this.size)]: this.size,
          [this.addPrefix('active')]: this.active,
          [this.addPrefix('disabled')]: this.disabled,
          [this.addPrefix('loading')]: this.loading,
          [this.addPrefix('block')]: this.block,
        },
      ]);
    },
  },

  methods: {
    _onClick() {
      this.$emit('click');
    },
  },
};
</script>
