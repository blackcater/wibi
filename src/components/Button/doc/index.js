export default [
  'basic',
  'appearance',
  'size',
  'color',
  'custom',
  'icon-button',
  'block',
  'disabled',
  'active',
  'loading',
  'group-basic',
  'vertical',
  'toolbar',
  'justified',
];
