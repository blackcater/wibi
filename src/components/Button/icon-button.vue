<template>
  <button/>
</template>

<script>
export default {
  name: 'IconButton',

  props: {},

  data() {
    return {};
  },

  methods: {},
};
</script>
