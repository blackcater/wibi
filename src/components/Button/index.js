import Button from './Button.jsx';
import ButtonGroup from './ButtonGroup.jsx';
import ButtonToolbar from './ButtonToolbar.jsx';
import IconButton from './IconButton.jsx';

Button.Icon = IconButton;
Button.Group = ButtonGroup;
Button.Toolbar = ButtonToolbar;

Button.install = function(Vue) {
  Vue.component(Button.name, Button);
  Vue.component(ButtonGroup.name, ButtonGroup);
  Vue.component(ButtonToolbar.name, ButtonToolbar);
};

export default Button;
