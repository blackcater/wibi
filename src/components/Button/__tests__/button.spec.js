import { mount } from '@vue/test-utils';
import Button from 'components/Button';

describe('Button component', () => {
  it('renders correctly', () => {
    const wrapper = mount({
      render() {
        return <Button>Default</Button>;
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
