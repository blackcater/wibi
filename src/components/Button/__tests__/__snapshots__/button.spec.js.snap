// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders correctly 1`] = `<button type="button" class="vs-btn vs-btn-default">Default<span class="vs-ripple-pond"><span class="vs-ripple" style="display: none;" enteractiveclass="vs-ripple-rippling"></span></span></button>`;
